src/controllers/scripting/controllerscriptenginebase.cpp

```diff
@@ -0,0 +1,251 @@
+#include "controllers/scripting/controllerscriptenginebase.h"
+
+#include "control/controlobject.h"
+#include "controllers/controller.h"
+#include "controllers/controllerdebug.h"
+#include "controllers/scripting/colormapperjsproxy.h"
+#include "errordialoghandler.h"
+#include "mixer/playermanager.h"
+
+ControllerScriptEngineBase::ControllerScriptEngineBase(Controller* controller)
+        : m_bDisplayingExceptionDialog(false),
+          m_pJSEngine(nullptr),
+          m_pController(controller),
+          m_bTesting(false) {
+    // Handle error dialog buttons
+    qRegisterMetaType<QMessageBox::StandardButton>("QMessageBox::StandardButton");
+
+    // Subclasses are responsible for adding paths to watch to m_scriptWatcher
+    // in their initializeScriptEngine method.
+    connect(&m_scriptWatcher,
+            &QFileSystemWatcher::fileChanged,
+            this,
+            &ControllerScriptEngineBase::reload);
+}
+
+ControllerScriptEngineBase::~ControllerScriptEngineBase() {
+    shutdown();
+}
+
+bool ControllerScriptEngineBase::initialize() {
+    VERIFY_OR_DEBUG_ASSERT(!m_pJSEngine) {
+        return false;
+    }
+
+    // Create the Script Engine
+    m_pJSEngine = new QJSEngine(this);
+
+    QJSValue engineGlobalObject = m_pJSEngine->globalObject();
+
+    QJSValue mapper = m_pJSEngine->newQMetaObject(
+            &ColorMapperJSProxy::staticMetaObject);
+    engineGlobalObject.setProperty("ColorMapper", mapper);
+
+    if (m_pController) {
+        qDebug() << "Controller in script engine is:"
+                 << m_pController->getName();
+
+        ControllerJSProxy* controllerProxy = m_pController->jsProxy();
+
+        // Make the Controller instance available to scripts
+        engineGlobalObject.setProperty(
+                "controller", m_pJSEngine->newQObject(controllerProxy));
+
+        // ...under the legacy name as well
+        engineGlobalObject.setProperty(
+                "midi", m_pJSEngine->newQObject(controllerProxy));
+    }
+
+    // Binary data is passed from the Controller as a QByteArray, which
+    // QJSEngine::toScriptValue converts to an ArrayBuffer in JavaScript.
+    // ArrayBuffer cannot be accessed with the [] operator in JS; it needs
+    // to be converted to a typed array (Uint8Array in this case) first.
+    // This function generates a wrapper function from a JS callback to do
+    // that conversion automatically.
+    m_makeArrayBufferWrapperFunction = m_pJSEngine->evaluate(QStringLiteral(
+            // arg2 is the timestamp for ControllerScriptModuleEngine.
+            // In ControllerScriptEngineLegacy it is the length of the array.
+            "(function(callback) {"
+            "    return function(arrayBuffer, arg2) {"
+            "        callback(new Uint8Array(arrayBuffer), arg2);"
+            "    };"
+            "})"));
+
+    return true;
+}
+
+void ControllerScriptEngineBase::shutdown() {
+    m_scriptWatcher.removePaths(m_scriptWatcher.directories());
```

Comment:
Adding and removing paths should be in one hand. 
I suggest to make m_scriptWatcher private and provide functions for that.

Appart from that this does not work, because we don't watch directories, do we?`

Reply:
What difference does it make to have this as a private member manipulated by protected functions?

Reply:
This will help to stream line the difficult QT API causing the bug here.

The more important part is to keep the responsibility for adding and removing a file in one hand.

Reply:
fixed in 3383aef and 19b9aaf